go/common/grpc/service.go

```diff
@@ -1,38 +1,132 @@
 package grpc
 
-import "fmt"
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/fxamacker/cbor"
+	"github.com/oasislabs/oasis-core/go/common"
+)
 
 // ServicePrefix is a prefix given to all gRPC services defined by oasis-core.
 const ServicePrefix = "oasis-core."
 
+var registeredMethods sync.Map
+
 // ServiceName is a gRPC service name.
 type ServiceName string
 
+// NamespaceExtractor extracts namespce from a method request.
+type NamespaceExtractor func(req interface{}) (common.Namespace, error)
+
+// ServiceNameFromMethod extract service name from method name.
+func ServiceNameFromMethod(methodName string) ServiceName {
+	substrs := strings.Split(methodName, "/")
+	// TODO: enforce no '/' in Service or Method names.
+	return ServiceName(substrs[1])
+}
+
 // NewServiceName creates a new gRPC service name.
 func NewServiceName(name string) ServiceName {
 	return ServiceName(ServicePrefix + name)
 }
 
-// NewMethodName creates a new method name for the given service.
-func (sn ServiceName) NewMethodName(name string) *MethodName {
-	return &MethodName{
-		short: name,
-		full:  fmt.Sprintf("/%s/%s", sn, name),
+// GetRegisteredMethod returns a registered method description.
+func GetRegisteredMethod(name string) (*MethodDesc, error) {
+	md, ok := registeredMethods.Load(name)
+	if !ok {
+		return nil, fmt.Errorf("method not registered")
+	}
+	m, ok := md.(*MethodDesc)
+	if !ok {
+		panic(fmt.Errorf("unexpected method description type: %T", md))
+	}
+	return m, nil
+}
+
+// NewMethodName is a temporray function until all service descriptions are migrated
+// to useing the NewMethodDesc method.
+func (sn ServiceName) NewMethodName(name string) *MethodDesc {
+	md := &MethodDesc{
+		short:       name,
+		full:        fmt.Sprintf("/%s/%s", sn, name),
+		requestType: nil,
+	}
+
+	if _, isRegistered := registeredMethods.Load(md.FullName()); isRegistered {
+		panic(fmt.Errorf("service: method already registered: %s", name))
+	}
+	registeredMethods.Store(md.FullName(), md)
+
+	return md
+}
+
+// NewMethodDesc creates a new method name for the given service.
+func (sn ServiceName) NewMethodDesc(name string, requestType interface{}) *MethodDesc {
+	md := &MethodDesc{
+		short:       name,
+		full:        fmt.Sprintf("/%s/%s", sn, name),
+		requestType: requestType,
 	}
+
+	if _, isRegistered := registeredMethods.Load(md.FullName()); isRegistered {
+		panic(fmt.Errorf("service: method already registered: %s", name))
+	}
+	registeredMethods.Store(md.FullName(), md)
+
+	return md
+}
+
+// WithNamespacePolicies tells weather the enpoint does have per namespace
+// polices defined.
+func (m *MethodDesc) WithNamespacePolicies(f NamespaceExtractor) *MethodDesc {
+	m.accessControl = true
+	m.namespaceExtractor = f
+
+	registeredMethods.Store(m.short, m)
```

Comment:
Hm this doesn't seem right? Shouldn't this be `m.full`?

Also are you sure you need to actually update the stored value? Since you are storing pointers, just updating `m` should be enough, assuming it is already registered?

Reply:
yeah you are right for both :+1: